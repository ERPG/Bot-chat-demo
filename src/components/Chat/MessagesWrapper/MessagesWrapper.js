import React, { useContext, useEffect, useState } from 'react';
import { GlobalLandbotContext } from '../../../contexts/LandbotContext';
import { messagesFilter, parseMessage, parseMessages, scrollBottom } from '../../../helpers/MessagesHelper';

import Message from '../../common/Message/Message';

import './MessageWrapper.css';

/** The MessagesWrapper is the container that loads all the messages between the user and the bot */

const MessagesWrapper = () => {
  const [messages, setMessages] = useState({});

  const Core = useContext(GlobalLandbotContext);

  useEffect(() => {
    if(Core) {
      subscribeToCore();
      initCore();
    }
  }, [Core]);

  const subscribeToCore = () => {
    Core.pipelines.$readableSequence.subscribe(data => {
      setMessages(messages => ({
        ...messages,
        [data.key]: parseMessage(data),
      }))
    });
  }

  const initCore = () => {
    Core.init().then(data => {
      setMessages(
        parseMessages(data.messages)
      );
    });
  }

  useEffect(() => {
    const container = document.getElementById('landbot-messages-container');
    scrollBottom(container);
  }, [messages]);

  return (
    <div id="landbot-messages-container" className="landbot-messages-container" data-landbot="message-container">
      {Object.values(messages)
        .filter(messagesFilter)
        .sort((a, b) => a.timestamp - b.timestamp)
        .map(message => (
            <Message key={message.key} author={message.author} text={message.text} />
        ))
      }
    </div>)
}

export default MessagesWrapper;