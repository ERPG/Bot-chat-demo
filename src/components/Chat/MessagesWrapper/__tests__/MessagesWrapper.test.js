import React from 'react';
import { render } from '@testing-library/react';
import { customRender } from '../../../../../__tests__/__utils__/testUtils';
import MessagesWrapper from '../MessagesWrapper';

describe('MessagesWrapper component', () => {
  const messageContainerSelector = 'message-container';

  beforeEach(() => {
    window.HTMLElement.prototype.scrollTo = jest.fn();
  });

  it('Should render component without errors', () => {
    const { container } = render(<MessagesWrapper />);
    expect(container).toMatchSnapshot();
  });

  it('Should have a container', () => {
    const { getByDataLb } = customRender(<MessagesWrapper />);
    const element = getByDataLb(messageContainerSelector)
    expect(element).toBeDefined();
  });
});