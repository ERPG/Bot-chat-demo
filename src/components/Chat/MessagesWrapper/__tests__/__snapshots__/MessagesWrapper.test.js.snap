// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesWrapper component Should render component without errors 1`] = `
<div>
  <div
    class="landbot-messages-container"
    data-landbot="message-container"
    id="landbot-messages-container"
  />
</div>
`;
